### PR TITLE
Remove misc_cast and validate types when casting

### PR DESCRIPTION
**Pull request by ouz-a**
Continuing our work in #102675

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_const_eval/src/transform/validate.rs [hunk 1]
```diff
@@ -564,13 +565,29 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
                     CastKind::PointerFromExposedAddress
                     | CastKind::PointerExposeAddress
                     | CastKind::Pointer(_) => {}
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/transform/validate.rs:567]: There are still some variants here that have no checks? Should have a FIXME at least. Also the comment above seems outdated.

File: compiler/rustc_const_eval/src/transform/validate.rs [hunk 2]
```diff
@@ -564,13 +565,29 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
...
+                            );
+                        }
+                    }
+                    CastKind::FnPtrToPtr | CastKind::PtrToPtr => {
+                        if !(op_ty.is_any_ptr() && target_type.is_unsafe_ptr()) {
+                            self.fail(location, "Can't cast {op_ty} into 'Ptr'");
+                        }
+                    }
+                    CastKind::FloatToFloat | CastKind::FloatToInt => {
+                        if !op_ty.is_floating_point() {
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/transform/validate.rs:586]: Should also check the target type.

File: compiler/rustc_const_eval/src/transform/validate.rs [hunk 3]
```diff
@@ -564,13 +565,29 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
                     CastKind::PointerFromExposedAddress
                     | CastKind::PointerExposeAddress
                     | CastKind::Pointer(_) => {}
-                    _ => {
-                        let op_ty = operand.ty(self.body, self.tcx);
-                        if op_ty.is_enum() {
+                    CastKind::IntToInt | CastKind::IntToFloat => {
+                        if !(op_ty.is_integral() || op_ty.is_char() || op_ty.is_bool())
+                            && !(target_type.is_floating_point()
```
[the remark this answers is not in the document]
Comment (ouz-a) [reply compiler/rustc_const_eval/src/transform/validate.rs]: Yeah that sounds super clean

File: compiler/rustc_const_eval/src/transform/validate.rs (hunk 1 above)
Comment (ouz-a) [reply compiler/rustc_const_eval/src/transform/validate.rs:567]: They were not part of the previous PR so I didn't touch em

File: compiler/rustc_const_eval/src/interpret/cast.rs [hunk 4]
```diff
@@ -143,18 +169,18 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 return Ok(self.cast_from_float(src.to_scalar().to_f64()?, cast_ty).into());
             }
             // The rest is integer/pointer-"like", including fn ptr casts
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/cast.rs:171]:
```suggestion
```

**Comment by RalfJung** [comment]
LGTM with that last comment fixed.

**Comment by oli-obk** [comment]
@bors delegate+

r=oli-obk,RalfJung with the commits squashed

**Comment by Dylan-DPC** [comment]
failed in [rollup](https://github.com/rust-lang/rust/pull/103391#issuecomment-1287655907) 
@bors r-

**Comment by RalfJung** [comment]
Ah, Miri needs to be patched as well. You can test this locally with `./x.py check src/tools/miri`.

**Pull request merged**